sdk/containerregistry/Azure.Containers.ContainerRegistry/src/Authentication/ContainerRegistryRefreshTokenCache.cs

```diff
@@ -0,0 +1,247 @@
+
+                // Access token has been successfully acquired in background and it is not expired yet, use it instead of current one
+                if (_backgroundUpdateTcs != null && _backgroundUpdateTcs.Task.Status == TaskStatus.RanToCompletion && _backgroundUpdateTcs.Task.Result.ExpiresOn > now)
+                {
+                    _infoTcs = _backgroundUpdateTcs;
+                    _backgroundUpdateTcs = default;
+                    return (_infoTcs, default, false);
+                }
+
+                // Attempt to get access token has failed or it has already expired. Need to get a new one
+                if (_infoTcs.Task.Status != TaskStatus.RanToCompletion || now >= _infoTcs.Task.Result.ExpiresOn)
+                {
+                    _infoTcs = new TaskCompletionSource<RefreshTokenInfo>(TaskCreationOptions.RunContinuationsAsynchronously);
+                    return (_infoTcs, default, true);
+                }
+
+                // Access token is still valid but is about to expire, try to get it in background
+                if (now >= _infoTcs.Task.Result.RefreshOn && _backgroundUpdateTcs == null)
+                {
+                    _backgroundUpdateTcs = new TaskCompletionSource<RefreshTokenInfo>(TaskCreationOptions.RunContinuationsAsynchronously);
+                    return (_infoTcs, _backgroundUpdateTcs, true);
+                }
+
+                // Access token is valid, use it
+                return (_infoTcs, default, false);
+            }
+        }
+
+        // must be called under lock (_syncObj)
+        private bool RequestRequiresNewToken(string service) =>
+            _currentTokenService == null ||
+            (service != null && !string.Equals(service, _currentTokenService));
+
+        private async ValueTask GetRefreshTokenFromCredentialInBackgroundAsync(TaskCompletionSource<RefreshTokenInfo> backgroundUpdateTcs, RefreshTokenInfo info, TokenRequestContext context, string service, bool async)
+        {
+            var cts = new CancellationTokenSource(_tokenRefreshRetryDelay);
+            try
+            {
+                RefreshTokenInfo newInfo = await GetRefreshTokenFromCredentialAsync(context, service, async, cts.Token).ConfigureAwait(false);
+                backgroundUpdateTcs.SetResult(newInfo);
+            }
+            catch (OperationCanceledException) when (cts.IsCancellationRequested)
+            {
+                backgroundUpdateTcs.SetResult(new RefreshTokenInfo(info.RefreshToken, info.ExpiresOn, DateTimeOffset.UtcNow));
+
+                // https://github.com/Azure/azure-sdk-for-net/issues/18539
+                //AzureCoreEventSource.Singleton.BackgroundRefreshFailed(context.ParentRequestId ?? string.Empty, oce.ToString());
+            }
+            catch (Exception)
+            {
+                backgroundUpdateTcs.SetResult(new RefreshTokenInfo(info.RefreshToken, info.ExpiresOn, DateTimeOffset.UtcNow + _tokenRefreshRetryDelay));
```

Comment:
Why do we add the tokenRefreshRetryDelay in the general exception case, but not for OperationCanceled?

Reply:
This is a @christothes question.

Reply:
I'm not sure either - in both cases we haven't actually updated the token result. Perhaps @AlexanderSher remembers.